### PR TITLE
2825 oxidation replace local manifest

### PR DESCRIPTION
**Pull request by FirelightFlagboy**
depends on #2728, depends on #2861

### CODE REVIEW

**Review by TimeEngineer — APPROVED** [review]

**Review by touilleMan — CHANGES_REQUESTED** [review]
🚧  review in progress 🚧

File: parsec/core/fs/remote_loader.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import trio
 from parsec._parsec import DateTime
 from trio import open_memory_channel, MemorySendChannel, MemoryReceiveChannel
+from parsec.api.data.manifest import ManifestTypeVar
```
Comment (touilleMan) [inline parsec/core/fs/remote_loader.py:9]: api data stuff should be acessed by doing

`frorm parsec.api.data import ManifestTypeVar, SequesterServiceID, ...`

Comment (touilleMan) [reply parsec/core/fs/remote_loader.py:9]: and parsec imports should be separated from 3rd party imports

File: parsec/core/types/manifest.py [hunk 2]
```diff
@@ -213,19 +112,17 @@ class LocalManifestType(Enum):
...
-    LocalFileManifestTypeVar,
-    LocalFolderManifestTypeVar,
-    LocalWorkspaceManifestTypeVar,
-    LocalUserManifestTypeVar,
+LocalManifestTypeVar = Union[
+    LocalFileManifest,
+    LocalFolderManifest,
+    LocalWorkspaceManifest,
+    LocalUserManifest,
 ]
```
Comment (touilleMan) [inline parsec/core/types/manifest.py:128]: Why do we still need `BaseLocalManifest` ?

If we keep using it, we still need all the python manifests classes, and end up mixing rust and python manifest classes (for instance creating a `parsec.api.dataFileManifest` creates a rust FileManifest, while deserializing with `parsec.api.data.BaseManifest.decrypt_verify_and_load` returns a python FileManifest...)

This can lead to subtle difference in behavior between the two classes that we'd better avoid

File: parsec/core/fs/workspacefs/sync_transactions.py [hunk 3]
```diff
@@ -224,7 +225,7 @@ def merge_manifests(
     assert remote_version > local_version and local_manifest.need_sync
 
     # All the local changes have been successfully uploaded
-    if local_manifest.match_remote(remote_manifest):
+    if local_manifest.match_remote(remote_manifest):  # type: ignore[arg-type]
```
Comment (touilleMan) [inline parsec/core/fs/workspacefs/sync_transactions.py:228]: why is this needed ? (and same question for the other type ignore in this file)

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 4]
```diff
@@ -630,7 +630,7 @@ async def _sync_by_id(
 
             # No new manifest to upload, the entry is synced!
             if new_remote_manifest is None:
-                return remote_manifest or (await self.local_storage.get_manifest(entry_id)).base
+                return remote_manifest or (await self.local_storage.get_manifest(entry_id)).base  # type: ignore[return-value]
```
Comment (touilleMan) [inline parsec/core/fs/workspacefs/workspacefs.py:633]: why the type ignore ?

File: parsec/core/fs/storage/manifest_storage.py [hunk 5]
```diff
@@ -308,7 +320,7 @@ def _thread_target() -> None:
             await self.localdb.run_in_thread(_thread_target)
 
         # Tag entry as up-to-date only if no new manifest has been written in the meantime
-        if manifest == self._cache[entry_id]:
+        if manifest == self._cache[entry_id]:  # type: ignore[operator]
```
Comment (touilleMan) [inline parsec/core/fs/storage/manifest_storage.py:323]: why ?

File: parsec/_parsec_pyi/local_manifest.pyi [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    def evolve_as_block(self, data: bytes) -> Chunk: ...
+    def is_block(self) -> bool: ...
+    def is_pseudo_block(self) -> bool: ...
+    def get_block_access(self) -> BlockAccess: ...
+    @classmethod
+    def from_block_access(cls, block_access: BlockAccess) -> Chunk: ...
+    @classmethod
+    def new(cls, start: int, stop: int) -> Chunk: ...
+
+class LocalFileManifest:
```
Comment (touilleMan) [inline parsec/_parsec_pyi/local_manifest.pyi:64]: We can cheat here by declaring the manifest inherits BaseManifest:
```python
from parsec.api.data import BaseManifest as BaseRemoteManifest
[...]
class FolderManifest(BaseRemoteManifest):
    [...]
```

This fine given it's only taken into account during typing (so zero risk this will break something at runtime)

File: parsec/core/fs/remote_loader.py [hunk 7]
```diff
@@ -637,14 +639,17 @@ async def load_manifest(
...
         try:
-            remote_manifest = BaseRemoteManifest.decrypt_verify_and_load(
-                rep["blob"],
-                key=workspace_entry.key,
-                author_verify_key=author.verify_key,
-                expected_author=expected_author,
-                expected_timestamp=expected_timestamp,
-                expected_version=expected_version,
-                expected_id=entry_id,
+            remote_manifest = cast(
```
Comment (touilleMan) [inline parsec/core/fs/remote_loader.py:642]: not needed with the "fake inheritance in .pyi" trick ^^

File: parsec/core/types/manifest.py [hunk 8]
```diff
@@ -54,9 +67,8 @@
 ChunkIDField: Type[fields.Field] = fields.uuid_based_field_factory(ChunkID)
 
 
-@functools.total_ordering
 @attr.s(slots=True, frozen=True, auto_attribs=True, kw_only=True, eq=False)
-class Chunk(BaseData):
+class _PyChunk(BaseData):
```
Comment (touilleMan) [inline parsec/core/types/manifest.py:71]: keeping the class but removing half of it implementation ? 🤔

File: src/local_manifest.rs [hunk 9]
```diff
@@ -0,0 +1,1544 @@
...
+    #[classmethod]
+    fn decrypt_and_load(_cls: &PyType, encrypted: &[u8], key: &SecretKey) -> PyResult<Self> {
+        match libparsec::client_types::LocalFileManifest::decrypt_and_load(encrypted, &key.0) {
+            Ok(x) => Ok(Self(x)),
+            Err(err) => Err(PyValueError::new_err(err)),
+        }
+    }
+
+    #[classmethod]
+    #[args(blocksize = "512 * 1024")]
```
Comment (touilleMan) [inline src/local_manifest.rs:393]: Should use `libparsec::api_types::DEFAULT_BLOCK_SIZE` instead

File: tests/core/fs/workspacefs/test_workspace_fs.py [hunk 10]
```diff
@@ -7,7 +7,7 @@
 from trio import open_nursery
 
 from parsec.api.protocol import DeviceID, RealmID, RealmRole
-from parsec.api.data import BaseManifest as BaseRemoteManifest, EntryName
```
Comment (touilleMan) [inline tests/core/fs/workspacefs/test_workspace_fs.py:10]: Strange renaming, especially given this class is monkeypatched in the test (where TypeVar are typing-only concepts)

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: parsec/core/fs/workspacefs/sync_transactions.py (hunk 3 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/workspacefs/sync_transactions.py:228]: This is need because both `local_manifest` & `remote_manifest` are unions, but `match_remote` only allow a subset of the union on `remote_manifest`, example:

`LocalFileManifest::match_remote` only take a `FileManifest` not a `FolderManifest` nor an `Union[FileManifest, FolderManifest]`

File: parsec/_parsec_pyi/local_manifest.pyi (hunk 6 above)
Comment (FirelightFlagboy) [reply parsec/_parsec_pyi/local_manifest.pyi:64]: As you said, it's a cheat. For me it's better to rework the python code to not rely on the inheritance of `BaseRemoteManifest`

File: src/local_manifest.rs (hunk 9 above)
Comment (FirelightFlagboy) [reply src/local_manifest.rs:393]: It should but, `args` require string, and the constant `DEFAULT_BLOCK_SIZE` isn't a string.

File: parsec/core/fs/remote_loader.py (hunk 7 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/remote_loader.py:642]: This discussion will continue in https://github.com/Scille/parsec-cloud/pull/2838#discussion_r948964509 then

File: parsec/core/fs/storage/manifest_storage.py (hunk 5 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/storage/manifest_storage.py:323]: same reason evoked here: https://github.com/Scille/parsec-cloud/pull/2838#discussion_r949044013

File: parsec/api/data/manifest.py [hunk 11]
```diff
@@ -71,6 +71,12 @@ def make_obj(self, data: Dict[str, Any]) -> "WorkspaceEntry":
 T = TypeVar("T")
 BaseAPISignedDataTypeVar = TypeVar("BaseAPISignedDataTypeVar", bound="BaseAPISignedData")
 BaseManifestTypeVar = TypeVar("BaseManifestTypeVar", bound="BaseManifest")
+ManifestTypeVar = Union[
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply parsec/api/data/manifest.py:74]: done

File: parsec/core/fs/remote_loader.py (hunk 1 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/remote_loader.py:9]: done

File: parsec/core/types/manifest.py (hunk 2 above)
Comment (FirelightFlagboy) [reply parsec/core/types/manifest.py:128]: `BaseLocalManifest` and `BaseManifest` will be removed in a latter PR

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 4 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/workspacefs/workspacefs.py:633]: The ignore was not needed, It was a bad typing of the returned value

File: src/local_manifest.rs (hunk 9 above)
Comment (touilleMan) [reply src/local_manifest.rs:393]: then keep the default at `None`, then use the constant in the function body

File: parsec/core/types/manifest.py (hunk 8 above)
Comment (FirelightFlagboy) [reply parsec/core/types/manifest.py]: `_Py*` classes no more

File: src/local_manifest.rs (hunk 9 above)
Comment (FirelightFlagboy) [reply src/local_manifest.rs]: done

File: parsec/api/data/manifest.py [hunk 12]
```diff
@@ -71,6 +71,12 @@ def make_obj(self, data: Dict[str, Any]) -> "WorkspaceEntry":
 T = TypeVar("T")
 BaseAPISignedDataTypeVar = TypeVar("BaseAPISignedDataTypeVar", bound="BaseAPISignedData")
 BaseManifestTypeVar = TypeVar("BaseManifestTypeVar", bound="BaseManifest")
+AnyManifest = Union[
```
Comment (touilleMan) [inline parsec/api/data/manifest.py:74]: `AnyManifest` should be defined in the .pyi

File: parsec/core/fs/remote_loader.py [hunk 13]
```diff
@@ -16,6 +17,7 @@
     BlockAccess,
     RealmRoleCertificate,
     BaseManifest as BaseRemoteManifest,
+    AnyManifest as RemoteAnyManifest,
```
Comment (touilleMan) [inline parsec/core/fs/remote_loader.py:20]: The local one is called `AnyLocalManifest`, so we should call this one `AnyRemoteManifest`

File: parsec/core/fs/workspacefs/sync_transactions.py [hunk 14]
```diff
@@ -7,17 +7,19 @@
 
 from parsec.api.protocol import DeviceID
 from parsec.core.core_events import CoreEvent
-from parsec.api.data import EntryNameTooLongError, BaseManifest as BaseRemoteManifest
+from parsec.api.data import EntryNameTooLongError, AnyManifest as RemoteAnyManifest
```
Comment (touilleMan) [inline parsec/core/fs/workspacefs/sync_transactions.py:10]: same as above

File: parsec/core/fs/workspacefs/workspacefs.py [hunk 15]
```diff
@@ -9,7 +9,7 @@
 from parsec.core.fs.workspacefs.entry_transactions import BlockInfo
 from parsec.crypto import CryptoError
 from parsec.event_bus import EventBus
-from parsec.api.data import BaseManifest as BaseRemoteManifest, BlockAccess
+from parsec.api.data import AnyManifest as RemoteAnyManifest, BlockAccess
```
Comment (touilleMan) [inline parsec/core/fs/workspacefs/workspacefs.py:12]: same as above

File: parsec/core/types/manifest.py [hunk 16]
```diff
@@ -54,994 +56,52 @@
...
 
-_PyLocalUserManifest = LocalUserManifest
-if not TYPE_CHECKING:
-    try:
-        from libparsec.types import LocalUserManifest as _RsLocalUserManifest
-    except:
-        pass
-    else:
-        LocalUserManifest = _RsLocalUserManifest
+def from_remote(
```
Comment (touilleMan) [inline parsec/core/types/manifest.py:74]: should call them `manifest_from_remote`/``manifest_from_remote_with_local_context`

File: tests/core/fs/workspacefs/test_workspace_fs.py (hunk 10 above)
Comment (touilleMan) [reply tests/core/fs/workspacefs/test_workspace_fs.py:10]: should revert this rename

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: tests/core/fs/workspacefs/test_workspace_fs.py (hunk 10 above)
Comment (FirelightFlagboy) [reply tests/core/fs/workspacefs/test_workspace_fs.py:10]: Next comment the complete change, for me you were commenting a deleted line

Comment (FirelightFlagboy) [reply tests/core/fs/workspacefs/test_workspace_fs.py]: done

File: parsec/api/data/manifest.py (hunk 12 above)
Comment (FirelightFlagboy) [reply parsec/api/data/manifest.py:74]: will be done in #2865

File: parsec/core/fs/remote_loader.py (hunk 13 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/remote_loader.py:20]: will be done in #2865

File: parsec/core/fs/workspacefs/sync_transactions.py (hunk 14 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/workspacefs/sync_transactions.py:10]: will be done in #2865

File: parsec/core/fs/workspacefs/workspacefs.py (hunk 15 above)
Comment (FirelightFlagboy) [reply parsec/core/fs/workspacefs/workspacefs.py:12]: will be done in #2865

File: parsec/core/types/manifest.py (hunk 16 above)
Comment (FirelightFlagboy) [reply parsec/core/types/manifest.py]: done

**Review by touilleMan — APPROVED** [review]
